Detect changes seems to not work if all changes are added to the index

russell opened the pull request:
`git add` adds the changed files to the index

Then `diff-files` diffs changes between the working tree and the index, there should be no changes if all the changes are correctly staged.

This PR changes the detect changes method so it instead diffs the contents of the index with the repository.  So if there are staged changes, it will detect them.

I have also exposed the repo and owner, since it's useful so we can update repositories other than ourselves

timbertson commented on action.yml:
```diff
@@ -11,6 +11,8 @@ inputs:
   prTitle: { "description": 'TODO', required: false }
   prBody: { "description": 'TODO', required: false }
+  repo: { "description": 'TODO', required: false }
+  owner: { "description": 'TODO', required: false }
```
This isn't used anywhere (yet)?

russell replied:
They are propagated through via the action constructor, but i think i can get away with setting `GITHUB_REPOSITORY` so this might not be needed.  I'll remove it and see

russell replied:
I tried passing this as part of the ENV but it was ignored

timbertson replied:
Ah, I was going to say "you'll need to add them to https://github.com/timbertson/self-update-action/blob/f417b06edcafd811cfe9129727f36fb74d13ce60/src/lib.ts#L31" but now I see that I already listed them, just forgot to actually include in `action.yml`. Got it 👍

timbertson commented:
> if there are staged changes, it will detect them.

Ignoring this is by design, so we should figure out if you're confused about the intent or whether you have a strange use case. The readme says:

> If the applyUpdateScript setting is provided:
> - Stage all changes from the updateScript step, so that they don't count as changes
> - Evaluate the applyUpdateScript setting

The initial use case is bumping a dependency for static manifest generation. If I make some change to bump `some-dependency` from `<old-commit>` -> `<current-master>`, then that will obviously be a change in git. But I only actually want to make a PR if the _result_ (my new generated manifests) have changed.

I've considered making this more extensible, so for example you could say "create a PR if the result of `dhall --file somefile.dhall` differs after the update". But I don't think that would remove the need to ignore staged changes in the simple case.

---

So, on to your use case. If you want all changes to be included, you should typically just provide an `updateScript` but no `applyUpdateScript`.

If it's something in your `updateScript` is already staging those changes, I guess you could unstage them at the end of your script? That's definitely a bit awkward, but I'd be surprised if that's commonplace.

russell commented:
> > if there are staged changes, it will detect them.
> 
> Ignoring this is by design, so we should figure out if you're confused about the intent or whether you have a strange use case. The readme says:
> 
> > If the applyUpdateScript setting is provided:
> > 
> > * Stage all changes from the updateScript step, so that they don't count as changes
> > * Evaluate the applyUpdateScript setting

I totally over looked this.

The problem I'm hitting is that `git commit --all` isn't sufficient to add new files.

```
       -a, --all
           Tell the command to automatically stage files that have been modified and deleted, but new files you have not told Git about are not affected.
```

So when the update script results in new files, then the update PR is missing new files which is a big limitation for us.  We are hoping to use this to update files as well as add new ones.

> The initial use case is bumping a dependency for static manifest generation. If I make some change to bump `some-dependency` from `<old-commit>` -> `<current-master>`, then that will obviously be a change in git. But I only actually want to make a PR if the _result_ (my new generated manifests) have changed.

I'm not sure why there are 2 update steps? i'm guessing it's because you want to avoid lock-files, but it makes more sense to me to unstage the lock files before diffing the index. They would all be added by the `--all` since they are already tracked.

> I've considered making this more extensible, so for example you could say "create a PR if the result of `dhall --file somefile.dhall` differs after the update". But I don't think that would remove the need to ignore staged changes in the simple case.
> 
> So, on to your use case. If you want all changes to be included, you should typically just provide an `updateScript` but no `applyUpdateScript`.
> 
> If it's something in your `updateScript` is already staging those changes, I guess you could unstage them at the end of your script? That's definitely a bit awkward, but I'd be surprised if that's commonplace.

It feels weird to me that by design this action only uses staging changes as a way to ignore what would be committed. If the update only results in new files, then the action as it currently stands will ignore the change entirely.

The way this currently works is that changes that aren't included in the index are used to determine if there was a change, this feels backwards, when I use git locally, the index is used to build up what will be committed, not the other way around.

I would propose what we change the action to behave like this

1. update script is run
2. stage changes (optional: default `git add .`)
3. detect changes (diff the index and the repo)
4. after detect changes script (optional: default `git add .`)
5. commit changes (commit changes exactly as they are in the index)

if you want to ignore some changes in the detect changes step, then you could declare that the script at step [2] was `git add . && git reset -- vendir.lock.yml` that would not result in the file being excluded from the commit, just from the detect changes step.  If this doesn't suffice, I would like to know more about your use case.

timbertson commented:
> So when the update script results in new files, then the update PR is missing new files which is a big limitation for us. We are hoping to use this to update files as well as add new ones.

That's totally reasonable, and I agree we should support it :+1:

## My use case

> I'm not sure why there are 2 update steps? i'm guessing it's because you want to avoid lock-files, but it makes more sense to me to unstage the lock files before diffing the index.

I'll give a concrete example, just so we're clear. Say I have a dhall file which imports https://github.com/regadas/github-actions-dhall:

```dhall
-- dependencies/Workflow.dhall
https://raw.githubusercontent.com/regadas/github-actions-dhall/dcd71670484439bfe7fe14ba669d6e137076a17b/schemas.dhall
```

My `updateScript` would be:

```
./dhall/bump --to regadas/github-actions-dhall dependencies/Workflow.dhall
```

Which results in this diff:

```diff
--- dependencies/Workflow.dhall
+++ dependencies/Workflow.dhall
@@ -1 +1 @@
-https://raw.githubusercontent.com/regadas/github-actions-dhall/dcd71670484439bfe7fe14ba669d6e137076a17b/schemas.dhall
+https://raw.githubusercontent.com/regadas/github-actions-dhall/3e03d6e9d38e8bef703201fabe33f3ad42316f6e/schemas.dhall
```

But do I want to make a PR for this? If it's just a change to the README or some dhall function I'm not using, then I would prefer not since it's just pointless noise. So my `applyUpdateScript` would be:

```
./script/render
```

This re-renders my `.github/workflows` files, using the latest version of github-actions-dhall. If the _render_ step produces a diff, then I do want to make the PR, since the update actually altered the generated files.

I am using the index slightly weird, in that it's a way to have changes on the filesystem which don't register as a change. But it works well for this use case. And it seems _somewhat_ general, since bumping some dependency will often change a git commit but that doesn't necessarily imply a change you actually care about.

## Your use case

Now let's talk about your use case. I have a couple observations:

 - You shouldn't have to care about my weird git index shenanigans because you're not even using the feature (`applyUpdateScript`) that it serves
 - The main thing you want is to track new files. You _probably_ don't care about the index too much, as long as it does this correctly?

I can't quite tell if your proposal would work, but even if it did it would make my workflow awkward (e.g. if my script is `make update`, that might update arbitrary dhall files so I don't know what to unstage). Let's see if we can both have our cake, shall we? I have two ideas:

#### make sure new files are counted as a diff:

After `updateScript`, we can run `git add --intent-to-add .`. This would mean that all new files would be included in `git diff` / `git commit --all`. I think that would fix your use case because now it tracks new files.

This solves your use case, but not the general one of "my updateScript staged some files". It's unclear if I need to support that, since the workaround is "just... don't do that" 😛 

#### include changes in the index when `applyUpdateScript` is not given

This would make the code a little more complex / hard to trace, since it acts differently in the two cases. But it would _definitely_ solve your problems since you don't actually want the "ignore some changes in the index" that the `applyUpdateScript` relies on.

----

I'm inclined to do the first, since it's pretty simple and I don't see any downsides. I don't see a need for the second yet, but I'm satisfied that it's at least possible if future use cases need it.

russell commented:
> I am using the index slightly weird, in that it's a way to have changes on the filesystem which don't register as a change. But it works well for this use case. And it seems somewhat general, since bumping some dependency will often change a git commit but that doesn't necessarily imply a change you actually care about.

I see what it's doing.  I don't doubt that it's possible. But I don't think it's really intuitive or obvious at the moment.  Instead of being explicit about what we want to ignore, we're using the order scripts are executed to determine what is included in the change. The purpose of the scripts isn't clear from the names, one script is called `applyUpdateScript` and the other `updateScript`, changes from `updateScript` are ignored (if you specify an `applyUpdateScript`).  It's not clear to me from the word `apply` that this behaviour change would occur

> I can't quite tell if your proposal would work, but even if it did it would make my workflow awkward (e.g. if my script is make update, that might update arbitrary dhall files so I don't know what to unstage). Let's see if we can both have our cake, shall we?

It does but it requires that you are explicit about what you want to ignore. Which is different to having one script's changes to the repository be ignored.  Instead it would require that if you wanted changes to be ignored that the developer specify what those changes are

I think that i'm approaching this from the perspective of a git workflow, and you're coming at it from this update use case.  As a result there is implicit behaviour in this action to reduce the need to be explicit the action on the users part,

> After updateScript, we can run git add --intent-to-add .. This would mean that all new files would be included in git diff / git commit --all. I think that would fix your use case because now it tracks new files.

That `git add --intent-to-add .` would need to be run after every script, not just the updateScript, also the afterUpdateScript, either script can result in new files.


---

I think the 2 things that would be nice to change in this action are:

the names of the `applyUpdateScript` and `updateScript` I would suggest the naming `setupScript` and `updateScript` where any change happening in the setup script are always ignored (for the purpose of detecting a change), and the setup script is optional.

Support for new files, not just existing one.  at the end of the day you can decide how you want to do that, it can be done with `--intent-to-add` for sure

timbertson commented:
> we're using the order scripts are executed to determine what is included in the change

Yep, by design. In any scenario where you're checking in generated files which are only _sometimes_ changing due to the update step, this seems ideal, because there will always be these two steps, in this order. There's no point to ignoring particular files by path, you want to ignore the result of the initial step.

If you're using vendir + jsonnet, I'm confused why you wouldn't have the same need. But maybe you're not committing the generated JSON, so you just _always_ bump the vendir'd code?

> the names of the applyUpdateScript and updateScript I would suggest the naming setupScript and updateScript where any change happening in the setup script are always ignored (for the purpose of detecting a change), and the setup script is optional.

Yeah, I think that makes sense. Making the first step the optional one incidentally means we only need to do the `--intent-to-add` in one place, too (since `setupScript` would now be followed by the blanket `git add .`). I'll probably need to do it on a `v1` branch, since it's a breaking change. Should have started with `v1`, not `main`...

russell commented:
I have refactored this PR to match what i think we are agreeing on as the interface, i'm going to merge all my outstanding PRs into a branch in the Zendesk repository so I can reference a somewhat stable ref.

russell committed:
Refactor to include a setup script

If declared the setupScript will ignore all changes that result from
running it.  They are ignored by adding them to the index

timbertson commented:
> I have refactored this PR to match what i think we are agreeing on as the interface, i'm going to merge all my outstanding PRs into a branch in the Zendesk repository so I can reference a somewhat stable ref.

Sounds good, thanks. It might take me another couple days to review / integrate, so if you're unblocked that's good. When I do merge I'll do it into a new `v1` branch, FYI

timbertson commented:
Merged into now-primary `v1` branch (929aa27c3ec23be34f99b1ba201b05b33e1aa8d1), I'm not sure why GH doesn't show it as merged.
